sky/adaptors/aws.py

```diff
@@ -41,8 +42,23 @@ def session():
     # Reference: https://github.com/boto/boto3/issues/1592
     # However, the session object itself is thread-safe, so we are
     # able to use lru_cache() to cache the session object.
-    with _session_creation_lock:
-        return boto3.session.Session()
+
+    # Retry 5 times by default for potential credential errors,
+    # mentioned in
+    # https://github.com/skypilot-org/skypilot/pull/1988
+    max_attempts = 5
+    attempt = 0
+    err = None
+    while attempt < max_attempts:
+        try:
+            with _session_creation_lock:
+                return boto3.session.Session()
+        except (botocore_exceptions().CredentialRetrievalError,
+                botocore_exceptions().NoCredentialsError) as e:
+            time.sleep(5)
+            err = e
+            attempt += 1
+    raise err
 
 
 @functools.lru_cache()
```

Comment:
(Like you discussed, this probably needs to go away as it's a memo table for all threads of the same process)

Can we refactor the code somehow so it's cached in a ThreadLocal?

Reply:
I find out the resource creation is actually very fast (<1s), so there is no need to cache it.